### DISCUSSION TITLE
로깅 정책 관련 논의

### DISCUSSION THREAD

**Discussion by Juhyung990122** [woowacourse-teams/2022-ternoko#207] [General]
터놓고 팀 내에서 사용할 로깅정책을 정리해보았습니다. 
저희 팀에서는 크게 "디버깅", "메소드 성능 측정", "사용자 접근 통계" 용도로 사용하려고 하는데요.
제안 드린 내용 외에도 활용할 방안이나, 수정해야할 것이 있다면  의견 남겨주세요.

# 로깅정책

~INFO :  스프링에서 제공하는 기본설정 사용

WARN :  요청 정보 찍고(INFO) + 우리가 핸들링하는 Exception 정보

ERROR : 요청정보 찍고(INFO)  + 우리가 핸들링하지 않는 Exception 정보

# PROD

- WARN 레벨까지 로깅
- CloudWatch 에서 추적하게 설정
- ERROR 레벨 발생시 Slack으로 메세지 전송
- 사용자 접근 통계가 필요한 기능 추적
    - 팀원 의견 필요(어떤 기능에 달건지)

# DEV

- INFO 레벨까지 로깅
- 서버 내 file로 저장(유효기간 2주 ~ 1달 사이 - 팀원 의견 필요)
    - 실시간으로 찍히는 파일(실시간 디버깅용(nohup 대체))
    - 시간단위로 잘려서 저장되는 파일(저장용)
- 성능 측정이 필요한 메소드 추적
    - 팀원 의견 필요(어떤 메소드에 달건지)

# LOCAL

- INFO 레벨까지 로깅 (팀원 의견 필요 - DEBUG 레벨로 할지)
- console 표시

**Comment by HyeonbinSa** [comment]
로깅 정책 관련해서 고민하시느라 고생 많았습니다. 깔끔하게 정리해주셔서 생각하기 수월했습니다.
### PROD
- CloudWatch 에서 추적하게 설정 (동의)
    - 클라우드 워치에서 추적한다면 파일로 DEV에서 설정 내용(콘솔 출력 및 파일 저장) + 클라우드 워치인지?
- ERROR 레벨 발생시 Slack으로 메세지 전송
    - CloudWatch를 사용한다면 경보 기능을 사용할 수 있을 것 같은데 Slack 알람 기능 또한 추가할지?
- 사용자 접근 통계가 필요한 기능 추적
   - 사용자 로그인 이력

### DEV
- 현재 서버 성능이나 관리 측면에서 로그를 2주 정도 가지고 있으면 충분할 것 같습니다.
- 성능 측정 같은 경우는 JPA를 활용해서 update하거나 select할 때 측정해보면 좋을 것 같습니다.

### LOCAL
- 로컬은 DEBUG 까지 봐도 좋을것같습니다. 계속해서 확인할 일이 많을 것 같아서...

**Reply by Juhyung990122** [reply]
안녕하세요 바니! 남겨주신 의견에 답변 드립니다.

# 질의응답
- 클라우드 워치에서 추적한다면 파일로 DEV에서 설정 내용(콘솔 출력 및 파일 저장) + 클라우드 워치인지?
아닙니다! 해보면서 변경될 것 같긴한데 PROD 로그 저장 및 관리는 전부 CloudWatch에 위임할 생각이에요.
PROD서버는 최대한 가볍게 두는게 좋을 것 같아서요.

- CloudWatch를 사용한다면 경보 기능을 사용할 수 있을 것 같은데 Slack 알람 기능 또한 추가할지?
저는 경보알림 이외에도 알람이 따로 있으면 좋겠다는 생각입니다. 왜냐하면 검색해보니 경보는 다음과 같은 형태로 오게 되는데요.
![image](https://user-images.githubusercontent.com/43205258/184519064-67ae8953-04b6-4526-8c4c-8bc066c08565.png)
보시다싶이 직관적으로 이해하기 어려워서 저희 팀이 장애원인을 파악하는데에 조금 어려울 것 같다는 생각이 들었습니다.
다만 이 이메일 내용을 제가 커스텀할 수 있다면 슬랙 알람이 크게 필요하지 않을 수도 있겠네요!
한번 조사해보고 슬랙 알람 도입 여부를 결정해보도록 하겠습니다ㅎㅎ

# Local 레벨
LOCAL 레벨은 DEBUG로 내리게 될 경우 밑에 앤지가 올려주신 사진처럼 정보가 뜨는데요!
저는 평소 저희가 보고 디버깅을 하는 레벨이 INFO 이고 DEBUG가 찍어주는 정보가 너무 많다고 생각되어서
현재 로그를 다음과 같이 개선해보았습니다.

![스크린샷 2022-08-14 오전 10 53 35](https://user-images.githubusercontent.com/43205258/184519369-887f11e6-b853-4cdb-927b-b0e60c291655.png)

이렇게 하면 INFO 레벨에서도 디버깅하기엔 충분하지 않을까 싶은데 바니 생각은 어떠신가요?
개선안 사진 보시구 한번 더 의견 부탁드립니다:)

이외의 내용은 동의합니다! 말씀주신 내용들 한번 반영해볼게요 의견 감사합니다!

**Comment by soominsohn** [comment]
안녕하세요 열음! 로그 분석 열심히 해주셨네요 감사합니다!

남겨주신 내용에 대해 제 의견 남기겠습니다. ㅎㅎ

INFO level로 돌려보고 일부러 예외 요청을 보낸 결과, 로그가 뜨고 있지 않습니다.

![image](https://user-images.githubusercontent.com/83059234/184265946-d016f897-2314-4501-b566-73f62d49abfd.png)

log level을 DEBUG로 해야 다음과 같은 로그를 확인 할 수 있었습니다.

![image](https://user-images.githubusercontent.com/83059234/184265962-d2c0d061-d206-4fef-ba25-79cd93c49b23.png)

제 의견은, PROD는 WARN level, DEV와 LOCAL은 DEBUG level로 가는게 어떨지 입니다.

### **PROD**

- WARN 레벨까지 로깅
    - 동의합니다.
- CloudWatch 에서 추적하게 설정 (동의)
    - 이 부분은 어떤건지 확인이 필요할 것 같아서
    - CloudWatch를 사용한다면 경보 기능을 사용할 수 있을 것 같은데 Slack 알람 기능 또한 추가할지?
        - 동의합니다.
- ERROR 레벨 발생시 Slack으로 메세지 전송
    - 필요한 기능이라고 생각이 듭니다! 동의합니다.
- 사용자 접근 통계가 필요한 기능 추적
    - 피드백 기능을 많이 사용하는지 확인

### **DEV**

- INFO 레벨까지 로깅
    - DEV서버이다 보니 테스트를 많이 할 것 같은데, 자세한 에외 메세지 확인을 위해 DEBUG 모드로 하는 것이 좋을 것 같습니다.
- 서버 내 file로 저장(유효기간 2주 ~ 1달 사이 - 팀원 의견 필요)
    - 실시간으로 찍히는 파일(실시간 디버깅용(nohup 대체)) → 동의
    - 시간단위로 잘려서 저장되는 파일(저장용) → 동의
    - 2주면 충분할 것 같습니다! 현재 상황에서 2주가 지난 로그를 확인할 일은 드물 것 같습니다.
- 성능 측정이 필요한 메소드 추적
    - 팀원 의견 필요(어떤 메소드에 달건지)
    - 연관관계 매핑이 많이 들어간 도메인에다가 넣으면 어떨까요?

### **LOCAL**

- INFO 레벨까지 로깅 (팀원 의견 필요 - DEBUG 레벨로 할지)
    - 이 역시 로그에서 예외메세지까지 확인하면 좋을 것 같아서 DEBUG모드가 좋을 것 같습니다.
- console 표시 → 동의합니당.

**Reply by Juhyung990122** [reply]
안녕하세요 앤지! 테스트까지 해주시다니.. 꼼꼼한 확인 감사합니다.
남겨주신 의견에 답변 드립니다!!

# Local 레벨
지금 보고계신 로그는 설정이 springframework상의 로그만 띄우는 것으로 설정되어 있기때문에 
저희가 프로젝트 내에서 처리한 예외는 보이지않습니다. 이에 현재는 다음과 같이 변경해두었는데요.
이렇게 하면 INFO레벨에서도 저희가 만든 로그를 띄울 수 있게됩니다. (+ 사진내의 stack trace는 info로 변경할지 고민중입니다..🤔)

![스크린샷 2022-08-14 오전 10 53 35](https://user-images.githubusercontent.com/43205258/184519311-0d858274-1e78-4f61-892f-5b02da9c16a5.png)

이에 저는 디버깅을 위해서라면 INFO 레벨도 충분하지 않을까 싶어요! 

이외의 다른 내용들은 전부 동의합니다:) 
LOCAL에서의 레벨 설정은 개선안 사진 확인하시고 한번 더 의견 부탁드립니다ㅎㅎ
언제나 감사합니다!
